chalk-ir/src/visit/boring_impls.rs

```diff
@@ -0,0 +1,341 @@
+//! This module contains "rote and uninteresting" impls of `Visit` for
+//! various types. In general, we prefer to derive `Visit`, but
+//! sometimes that doesn't work for whatever reason.
+//!
+//! The more interesting impls of `Visit` remain in the `visit` module.
+
+use crate::{
+    AssocTypeId, DebruijnIndex, Goals, ImplId, Interner, Parameter, ParameterKind,
+    PlaceholderIndex, ProgramClause, QuantifierKind, StructId, Substitution, SuperVisit, TraitId,
+    UniverseIndex, Visit, VisitResult, Visitor,
+};
+use chalk_engine::{context::Context, ExClause, FlounderedSubgoal, Literal};
+use std::{marker::PhantomData, sync::Arc};
+
+impl<T: Visit<I>, I: Interner> Visit<I> for &T {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        (**self).visit_with(visitor, outer_binder)
```

Comment:
Nit: I'm not sure what we wrote elsewhere, but I've migrated to writing code like this as `T::visit_with(self, visitor, outer_binder)`. This removes the need to have a bunch of `**`, which I always find kind of hard to mentally track, and makes it quite clear that we are invoking the `Visitor` defined for the referent type `T`

Reply:
(I've had e.g. infinite loops because I wrote `(*self).` etc...there is of course a litn for that now, but the `T::` form seems more clear and secure.)